### ISSUE TITLE
ESP32: Prototype Open-Fleet API Concept With Golioth

### ISSUE THREAD

**Issue by keenanjohnson** [Ribbit-Network/ribbit-frog-micropython#6]

**Comment by keenanjohnson** [comment]
One big decision we will need to make is whether we are going to store user profiles for people building sensors? We haven't historically done that. Here are my thoughts

Pros:
* We can ensure that people are limited to one ID per device. If they have to re-register their device for some reason, they can get the same device ID, which will help limit the number of devices in the fleet. This is a problem with balena today as there are many "dead" devices in the fleet that were flashed once, then someone reflashed their device and got a new ID
* If we could associate an email or contact info with a device, we would have a way to reach out if say the sensor goes offline for some reason etc,

Cons
* If we have user data and authentication, we will have to ensure that we are good stewards of that data in terms of security and collection and privacy. 

The pros would be nice imo, though the security / data stewardship challenges trend me to believe that we should try to use something out of the box to limit our surface area for data leakage, etc. What do you all think?

**Comment by beriberikix** [comment]
If you decide to support and/or require login, I'd advise outsourcing it to a trusted third party like Auth0 and Firebase (there are many.) They can greatly reduce the risk of leaks, PII concerns, GDPR, etc.

In general it might be a good idea, not just for service communication but also managing abuse. Anonymous API users could be a pretty scary vector.

**Comment by abhinavtripathy** [comment]
@keenanjohnson my thoughts so far: 
- From what i have read an auth service would be ideal to use to avoid situations like DDos and perhaps having user information could prove useful in the future. 
- I would propose firebase/supabase as an auth service. I like supabase more since it is open sourced. The auth service we use is not a one way door its a 2 way door so it would be fairly easy to switch it to a different service/provider so I don't want to spend a lot of time thinking about what to choose, lets pick one and go ahead with it. 
- What do people think on using React for frend end and python or nodeJS for backend. Trying to keep it simple here. 
- The web flow makes sense, I might need help from folks in regards to what exactly the data is being sent (I don't have a lot of context to the frog sensor but I will skim through this thread again to get a better sense of what we are trying to accomplish)

Asks from you @keenanjohnson :
- Let us finalize an auth service and frameworks we should go ahead with
- Once an auth service is finalized we need an account I can run some tests on and enable permissions etc.

**Comment by abhinavtripathy** [comment]
in terms of an ECD, I am still working towards that. I will have a design doc shortly as well.

**Comment by damz** [comment]
Firebase sounds reasonable to me, as we are likely going to rely on other Google services (i.e. Pubsub and BigQuery)... but it is still a lot of reinventing the wheel. If we go with something like Django for the backend we get user authentication for free.

(Additionally, if we go with Django we don't have to go through the hassle of building separate backend and frontend applications. At the cost of sticking with an old-style backend driven frontend... but do we really need anything else?)

**Comment by abhinavtripathy** [comment]
Fair point, we can ahead with firebase. 

> but it is still a lot of reinventing the wheel.

not sure I understood what you mean by that, could you please explain more. 

> If we go with something like Django for the backend we get user authentication for free.

True, though I am not a huge fan of Django simple because it old style. Is the auth cost a big consideration for this? I propose React on the client side and we can go ahead with firebase functions or some other serverless options for the backend.

**Comment by mlbenthall** [comment]
It seems like it would be really nice to:
- Have the device registration tied to the physical device, not the instance of registration
- Not have to recompile the device software in order to insert a key

Is there any identifier that the sensor/ESP32 can instrospect about itself, that we could use as its key, or construct its key from? 
We'd need it to be something "unique". Also something that the user can discover and type in when registering.

**Comment by beriberikix** [comment]
Yup, every ESP32 (and just about every Wi-Fi device) has a unique ID based off it's base MAC address: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/system/misc_system_api.html#mac-address

For future devices that are not wi-fi, like cellular, also have a unique ID like IMEI.

**Comment by keenanjohnson** [comment]
Great points @mlbenthall and @beriberikix !

The software keys are modifiedable at runtime. [You can see the instructions for how to set them here.](https://ribbit-network.github.io/ribbit-network-frog-hardware/3-software/3-software/)

I think having the devices auto-provision could be a nice future goal, but I still think having the registration page would be valuable so that we can collect a contact email associated with the sensor to do future things like notify them if the sensor goes offline, etc.
